source/adapters/cuda/enqueue.cpp

```diff
@@ -565,45 +618,61 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueMemBufferReadRect(
     size_t hostRowPitch, size_t hostSlicePitch, void *pDst,
     uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
     ur_event_handle_t *phEvent) {
-  ur_result_t Result = UR_RESULT_SUCCESS;
-  CUdeviceptr DevPtr = std::get<BufferMem>(hBuffer->Mem).get();
   std::unique_ptr<ur_event_handle_t_> RetImplEvent{nullptr};
 
-  try {
-    ScopedContext Active(hQueue->getContext());
-    CUstream CuStream = hQueue->getNextTransferStream();
+  ur_lock MemoryMigrationLock{hBuffer->MemoryMigrationMutex};
+  auto Device = hQueue->getDevice();
+  CUstream Stream = hQueue->getNextTransferStream();
 
-    Result = enqueueEventsWait(hQueue, CuStream, numEventsInWaitList,
-                               phEventWaitList);
+  try {
+    // Note that this entry point may be called on a queue that may not be the
+    // last queue to write to the MemBuffer, meaning we must perform the copy
+    // from a different device
+    if (hBuffer->LastEventWritingToMemObj &&
+        hBuffer->LastDeviceWritingToMemObj != hQueue->getDevice()) {
+      Device = hBuffer->LastDeviceWritingToMemObj;
+      ScopedContext Active(Device);
+      Stream = CUstream{0}; // Default stream for different device
+      // We may have to wait for an event on another queue if it is the last
+      // event writing to mem obj
+      UR_CHECK_ERROR(enqueueEventsWait(hQueue, Stream, 1,
+                                       &hBuffer->LastEventWritingToMemObj));
+    }
+
+    ScopedContext Active(Device);
+
+    UR_CHECK_ERROR(enqueueEventsWait(hQueue, Stream, numEventsInWaitList,
+                                     phEventWaitList));
 
     if (phEvent) {
       RetImplEvent =
           std::unique_ptr<ur_event_handle_t_>(ur_event_handle_t_::makeNative(
-              UR_COMMAND_MEM_BUFFER_READ_RECT, hQueue, CuStream));
+              UR_COMMAND_MEM_BUFFER_READ_RECT, hQueue, Stream));
       UR_CHECK_ERROR(RetImplEvent->start());
     }
 
-    Result = commonEnqueueMemBufferCopyRect(
-        CuStream, region, &DevPtr, CU_MEMORYTYPE_DEVICE, bufferOrigin,
+    void *DevPtr = std::get<BufferMem>(hBuffer->Mem).getVoid(Device);
+    UR_CHECK_ERROR(commonEnqueueMemBufferCopyRect(
+        Stream, region, DevPtr, CU_MEMORYTYPE_DEVICE, bufferOrigin,
         bufferRowPitch, bufferSlicePitch, pDst, CU_MEMORYTYPE_HOST, hostOrigin,
-        hostRowPitch, hostSlicePitch);
+        hostRowPitch, hostSlicePitch));
 
     if (phEvent) {
       UR_CHECK_ERROR(RetImplEvent->record());
     }
 
     if (blockingRead) {
-      UR_CHECK_ERROR(cuStreamSynchronize(CuStream));
+      UR_CHECK_ERROR(cuStreamSynchronize(Stream));
```

Comment:
It seems like blockingRead is set false in dpc++ memory manager? https://github.com/intel/llvm/blob/c74cebdb19b5cd99f66afe781cd3874e88f66308/sycl/source/detail/memory_manager.cpp#L613

I didn't look thoroughly and I'm far from certain but to me it looks like possibly this change will break buffer memory consistency if you have two devices, 
- one writing to the buffer on device at same time as..
- other copying with Stream to host for this read?

If something like this is the case then I think correctness should be prioritized over performance: whatever happens unless/until until p2p is somehow enabled for multidevice buffers performance is going to be worse than using usm and P2P manually, which also in any case seems to be the priority for the majority of users. I think we should prioritize fully supporting multi device functionality with usm that this patch will enable, then having buffer multi device performance as a secondary thing. Although ideally it wouldn't make multi device buffer code any slower than the existing routes that does D2H H2D copies via the existing one device in one sycl context state of affairs.

Reply:
This patch uses locks and events to ensure that a buffer cannot be read from at the same time as an operation is writing to it. If I want to read from a buffer I need to wait until the last event writing to that buffer has completed, and then a memcpy will happen from the device where the buffer was last written to.

Reply:
See here for a more detailed explanation https://github.com/oneapi-src/unified-runtime/pull/1565/files#diff-28b062fd8949b104c030051b807123eb1f86396e56d7524e1fdae76187d02a70R335

Reply:
OK, it's not obvious to me how the lock works the prevent the above hypothetical issue in all cases since these operations are asynchronous, but I'll take your word for it: I can see that you wait on the stream that does the read, so I guess that you've made it such that all other operations acting on the buffer memory  for all other hypothetical devices also wait for stream completion.
For the reasons listed above regarding performance priorities I think this is all fine IMO.

Reply:
The operations are not asynchronous, the work that the operations enqueue is async. Use of locks like this enables safety in a multi threaded environment.

Reply:
Forgetting about multithreaded usage which I think is so niche in gpu projects outside of MPI (which is multi process rather than multithread within a single process so not affected by these locks anyway) to make it pretty much irrelevant.

My question is what happens if a single host thread
- 1. operates on buffer memory "mem" on device A using an async operation using e.g. stream 1 on device A
- 2. Following this it operates on buffer memory "mem" to copy it to device B using e.g. stream 2 on device B

I believe Sycl programming model mandates that the memory device B sees is consistent with the memory after the async operation on device A. Does the sycl scheduler still manage this dependency and enforce this memory consistency constraint? Or is it now the responsibility of the plugin? If it is the responsibility of the plugin then does it guarantee correctness of the example I gave (and all other possible combinations of operations)?

Reply:
Each `ur_mem_handle_t` has a `ur_event_handle_t LastEventWritingToMemObj`. If an operation writes to mem A, then this event will be set. When another operation wants to read from this mem A, either on the same or another device, then the plugin will ensure that the new command doesn't start (if same device)/isn't enqueued (for different device) until `LastEventWritingToMemObj` completes.

In the case of a single device ctx we just use the normal scheduling with streams to ensure the next enqueued operation only starts after `LastEventWritingToMemObj` completes. If we need to access data on a different device in the ctx, then the enqueue of the new command will need to wait on `LastEventWritingToMemObj` to complete.

Reply:
Great thanks, it all makes sense to me. Just wanted to make sure.